crates/runtime/src/instance/allocator/pooling/index_allocator.rs

```diff
@@ -0,0 +1,408 @@
+                debug_assert!(free_list.len() > 0);
+                free_list.pop().unwrap()
+            }
+            &mut PoolingAllocationState::Random(ref mut free_list) => {
+                debug_assert!(free_list.len() > 0);
+                let id = rand::thread_rng().gen_range(0..free_list.len());
+                free_list.swap_remove(id)
+            }
+            &mut PoolingAllocationState::ReuseAffinity {
+                ref mut free_list,
+                ref mut per_module,
+                ref mut slot_state,
+                ..
+            } => {
+                if let Some(this_module) = id.and_then(|id| per_module.get_mut(&id)) {
+                    // There is a freelist of slots with affinity for
+                    // the requested module-ID. Pick the last one; any
+                    // will do, no need for randomness here.
+                    assert!(!this_module.is_empty());
+                    let new_id = this_module.pop().expect("List should never be empty");
+                    if this_module.is_empty() {
+                        per_module.remove(&id.unwrap());
+                    }
+                    // Make sure to remove from the global
+                    // freelist. We already removed from the
+                    // per-module list above.
+                    Self::remove_free_list_item(slot_state, free_list, new_id);
+                    slot_state[new_id] = SlotState::Taken;
+                    new_id
+                } else {
+                    // Pick a random free slot ID. Note that we do
+                    // this, rather than pick a victim module first,
+                    // to maintain an unbiased stealing distribution:
+                    // we want the likelihood of our taking a slot
+                    // from some other module's freelist to be
+                    // proportional to that module's freelist
+                    // length. Or in other words, every *slot* should
+                    // be equally likely to be stolen. The
+                    // alternative, where we pick the victim module
+                    // freelist first, means that either a module with
+                    // an affinity freelist of one slot has the same
+                    // chances of losing that slot as one with a
+                    // hundred slots; or else we need a weighted
+                    // random choice among modules, which is just as
+                    // complex as this process.
+                    //
+                    // We don't bother picking an empty slot (no
+                    // established affinity) before a random slot,
+                    // because this is more complex, and in the steady
+                    // state, all slots will see at least one
+                    // instantiation very quickly, so there will never
+                    // (past an initial phase) be a slot with no
+                    // affinity.
```

Comment:
This is a very helpful comment, thanks for writing this!

Reply:
:+1: